sql/src/main/java/io/cloudevents/sql/impl/runtime/EvaluationResult.java

```diff
@@ -2,20 +2,79 @@
 
 import io.cloudevents.sql.EvaluationException;
 import io.cloudevents.sql.Result;
+import io.cloudevents.sql.Type;
 
-import java.util.Collection;
-import java.util.Collections;
-import java.util.List;
-import java.util.Objects;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 public class EvaluationResult implements Result {
 
     private final Object value;
     private final List<EvaluationException> exceptions;
+    private final EvaluationException latestException;
 
     public EvaluationResult(Object value, List<EvaluationException> exceptions) {
         this.value = value;
-        this.exceptions = exceptions == null ? Collections.emptyList() : Collections.unmodifiableList(exceptions);
+        this.exceptions = exceptions == null ? new ArrayList<>() : exceptions;
+        this.latestException = null;
+    }
+
+    public EvaluationResult(Object value, EvaluationException exception, EvaluationResult left, EvaluationResult right) {
+        this.exceptions = Stream.concat(Stream.of(left, right).filter(Objects::nonNull).map(r -> r.exceptions).flatMap(Collection::stream), Stream.of(exception))
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+        this.latestException = exception;
+        this.value = value;
+    }
+
+    public EvaluationResult(Object value, EvaluationException exception) {
+        this(value, exception, null, null);
+    }
+
+    public EvaluationResult(Object value) {
+        this.value = value;
+        this.exceptions = new ArrayList<>();
+        this.latestException = null;
+    }
+
+    public EvaluationResult wrap(EvaluationResult other) {
```

Comment:
Should we call this "wrapExceptions" as I believe it would make callers code a little more explicit on what are we wrapping

Reply:
including the overloaded one